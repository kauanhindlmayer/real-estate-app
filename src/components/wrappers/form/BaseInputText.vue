<script lang="ts" setup>
import InputText from 'primevue/inputtext'
import { useUid } from '@/composables/useUid'

defineProps<{ label?: string; error?: string }>()

const modelValue = defineModel<String | Number>()
const uid = useUid()
</script>

<template>
  <div class="field">
    <label v-if="label" :for="uid">
      {{ label }}
      <span v-if="required" class="p-error"> *</span>
    </label>
    <InputText
      v-bind="$attrs"
      v-model="modelValue"
      :id="uid"
      :aria-labelledby="error ? `${uid}-error` : undefined"
      :aria-invalid="!!error"
      :invalid="!!error"
      class="w-full"
    />
    <small class="p-error" v-if="error" :id="`${uid}-error`" aria-live="assertive">
      {{ error }}
    </small>
  </div>
</template>
