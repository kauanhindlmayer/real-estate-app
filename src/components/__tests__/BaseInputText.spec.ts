import { describe, it, expect, beforeEach } from 'vitest'
import { VueWrapper, mount } from '@vue/test-utils'

import PrimeVue from 'primevue/config'
import BaseInputText from '@/components/wrappers/form/BaseInputText.vue'

describe('BaseInputText', () => {
  let wrapper: VueWrapper<any>

  beforeEach(() => {
    wrapper = mount(BaseInputText, {
      global: {
        plugins: [PrimeVue]
      }
    })
  })

  it('renders properly', async () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders the label when provided', async () => {
    wrapper.setProps({ label: 'Name' })
    await wrapper.vm.$nextTick()
    const labelElement = wrapper.find('label')
    expect(labelElement.exists()).toBe(true)
    expect(labelElement.text()).toBe('Name')
  })

  it('does not render the label when not provided', async () => {
    const labelElement = wrapper.find('label')
    expect(labelElement.exists()).toBe(false)
  })

  it('renders the error message when input is invalid', async () => {
    wrapper.setProps({ modelValue: '', label: 'Name', error: 'This field is required' })
    await wrapper.vm.$nextTick()
    const errorMessageElement = wrapper.find('small.p-error')
    expect(errorMessageElement.exists()).toBe(true)
    expect(errorMessageElement.text()).toBe('This field is required')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('does not render the error message when input is valid', async () => {
    wrapper.setProps({ modelValue: 'John Doe' })
    await wrapper.vm.$nextTick()
    const errorMessageElement = wrapper.find('small.p-error')
    expect(errorMessageElement.exists()).toBe(false)
  })
})
