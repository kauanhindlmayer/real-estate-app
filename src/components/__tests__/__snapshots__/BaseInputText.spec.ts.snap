// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BaseInputText > renders properly 1`] = `
"<div class="field">
  <!--v-if--><input class="p-inputtext p-component w-full" aria-invalid="false" id="0" data-pc-name="inputtext" data-pc-section="root">
  <!--v-if-->
</div>"
`;

exports[`BaseInputText > renders the error message when input is invalid 1`] = `"<div class="field"><label for="3">Name <span class="p-error"> *</span></label><input class="p-inputtext p-component w-full" aria-invalid="true" id="3" data-pc-name="inputtext" data-pc-section="root" aria-labelledby="3-error"><small class="p-error" id="3-error" aria-live="assertive">This field is required</small></div>"`;
